File: flutter/packages/hatake_material/test/wizard_step_condition_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:hatake_material/hatake_material.dart';

/// 条件で丸ごと飛ばすステップ（`steps[].visibleWhen`）。
///
/// 守るのは**送り側と検証側が同じ答えを持つ**こと。片方だけだと、誰にも見えない
/// ステップの必須で保存できない画面（画面にはどこが悪いか出ない）か、条件が効いて
/// いない画面のどちらかになる。
class _Repo implements Repository {
  DataRecord? saved;

  @override
  Future<PageResult> search(RepositoryQuery query) async => PageResult.empty;
  @override
  Future<DataRecord?> findByKey(Object key) async => null;
  @override
  Future<DataRecord> create(DataRecord data) async => saved = data;
  @override
  Future<DataRecord> update(Object key, DataRecord data) async => saved = data;
  @override
  Future<void> delete(Object key) async {}
}

/// 区分が「法人」のときだけ請求先のステップを出すウィザード。
const _definition = WizardPageDefinition(
  id: 'customer_wizard',
  title: '顧客登録',
  repository: 'repo',
  steps: [
    WizardStepDefinition(
      id: 'basic',
      title: '基本情報',
      fields: [
        FieldDefinition(
          field: 'kind',
          label: '区分',
          type: FieldTypes.select,
          required: true,
          options: [
            OptionItem(value: 'corp', label: '法人'),
            OptionItem(value: 'person', label: '個人'),
          ],
        ),
      ],
    ),
    WizardStepDefinition(
      id: 'billing',
      title: '請求先',
      // 法人のときだけ。個人なら丸ごと飛ばす（必須も効かない）。
      visibleWhen: {'field': 'kind', 'operator': 'equals', 'value': 'corp'},
      fields: [
        FieldDefinition(field: 'billTo', label: '請求先', required: true),
      ],
    ),
    WizardStepDefinition(
      id: 'confirm',
      title: '確認',
      fields: [FieldDefinition(field: 'memo', label: '備考')],
    ),
  ],
);

Widget _harness(_Repo repo, {WizardPageDefinition definition = _definition}) =>
    MaterialApp(
      home: Scaffold(
        body: HatakeScope(
          repositories: RepositoryRegistry({'repo': repo}),
          renderer: const MaterialRenderer(),
          child: HatakeWizardView(definition: definition),
        ),
      ),
    );

/// 選択肢を選ぶ（開いてから、その札を押す）。
Future<void> _pick(WidgetTester tester, String label) async {
  await tester.tap(find.byKey(Key(HatakeKeys.form('kind'))));
  await tester.pumpAndSettle();
  await tester.tap(find.text(label).last);
  await tester.pumpAndSettle();
}

Future<void> _next(WidgetTester tester) async {
  await tester.tap(find.byKey(const Key(HatakeKeys.wizardNext)));
  await tester.pumpAndSettle();
}

void main() {
  testWidgets('条件に合うときは、そのステップを通る', (tester) async {
    await tester.pumpWidget(_harness(_Repo()));
    await tester.pumpAndSettle();

    await _pick(tester, '法人');
    await _next(tester);

    expect(find.byKey(Key(HatakeKeys.form('billTo'))), findsOneWidget);
    // 歩数の表示にも出る。
    expect(find.byKey(Key(HatakeKeys.wizardStep('billing'))), findsOneWidget);
  });

  testWidgets('**条件に合わないステップは丸ごと飛ばす**（歩数の表示にも出さない）',
      (tester) async {
    await tester.pumpWidget(_harness(_Repo()));
    await tester.pumpAndSettle();

    await _pick(tester, '個人');
    await _next(tester);

    // 請求先は出ない＝次は確認。
    expect(find.byKey(Key(HatakeKeys.form('billTo'))), findsNothing);
    expect(find.byKey(Key(HatakeKeys.form('memo'))), findsOneWidget);
    expect(find.byKey(Key(HatakeKeys.wizardStep('billing'))), findsNothing);
    // 飛ばした先が最後なら「保存」になる（次へは出ない）。
    expect(find.byKey(const Key(HatakeKeys.wizardSave)), findsOneWidget);
  });

  testWidgets('**隠れたステップの必須は保存を止めない**（検証も飛ぶ）', (tester) async {
    final repo = _Repo();
    await tester.pumpWidget(_harness(repo));
    await tester.pumpAndSettle();

    await _pick(tester, '個人');
    await _next(tester);
    await tester.tap(find.byKey(const Key(HatakeKeys.wizardSave)));
    await tester.pumpAndSettle();

    // billTo は必須だが、そのステップは誰にも見えていない＝保存できる。
    expect(repo.saved, isNotNull);
    expect(repo.saved!['kind'], 'person');
    expect(repo.saved!.containsKey('billTo'), isFalse);
  });

  testWidgets('戻るときも飛ばす（送り側と検証側で答えが違わない）', (tester) async {
    await tester.pumpWidget(_harness(_Repo()));
    await tester.pumpAndSettle();

    await _pick(tester, '個人');
    await _next(tester);
    expect(find.byKey(Key(HatakeKeys.form('memo'))), findsOneWidget);

    await tester.tap(find.byKey(const Key(HatakeKeys.wizardBack)));
    await tester.pumpAndSettle();

    // 1つ戻ると、請求先ではなく基本情報。
    expect(find.byKey(Key(HatakeKeys.form('kind'))), findsOneWidget);
    expect(find.byKey(Key(HatakeKeys.form('billTo'))), findsNothing);
  });

  testWidgets('全部隠れたら、黙って1枚目を出さない', (tester) async {
    const nothingShown = WizardPageDefinition(
      id: 'w',
      title: '出せないウィザード',
      repository: 'repo',
      steps: [
        WizardStepDefinition(
          id: 'only',
          title: '唯一',
          visibleWhen: {'field': 'kind', 'operator': 'equals', 'value': 'never'},
          fields: [FieldDefinition(field: 'a', label: 'A')],
        ),
      ],
    );
    await tester.pumpWidget(_harness(_Repo(), definition: nothingShown));
    await tester.pumpAndSettle();

    expect(find.byKey(const Key(HatakeKeys.wizardEmpty)), findsOneWidget);
    expect(find.byKey(Key(HatakeKeys.form('a'))), findsNothing);
  });
}

File: typescript/test/wizardStepCondition.test.ts
import { describe, expect, it } from "vitest";
import {
  explainSource,
  FormValidator,
  isStepVisible,
  parsePageYaml,
  UnknownKeysError,
  visibleWizardSteps,
  wizardForm,
  wizardStepForm,
  type WizardPageDefinition,
} from "../src/index.js";

/**
 * ウィザードのステップ条件（`steps[].visibleWhen`）。
 *
 * 条件はどこにでも書けるのに**ステップだけ書けなかった**（語彙の穴）。埋めるときに
 * 守るのは「**送り側と検証側が同じ答えを持つ**」こと＝誰にも見えないステップの必須で
 * 保存できない画面を作らない。そのために、ステップは区画に開いて条件も一緒に渡す
 * （「隠れた区画は検証しない」を使い回す＝判定を2つ持たない）。
 */
const YAML = `dsl_version: "1.0"

page:
  type: wizard
  id: customer_wizard
  title: 顧客登録
  repository: customerRepository
  key: id
  steps:
    - id: basic
      title: 基本情報
      fields:
        - { field: kind, label: 区分, type: select, required: true,
            options: [{ value: corp, label: 法人 }, { value: person, label: 個人 }] }
    - id: billing
      title: 請求先
      visibleWhen: { field: kind, operator: equals, value: corp }
      fields:
        - { field: billTo, label: 請求先, required: true }
    - id: confirm
      title: 確認
      fields:
        - { field: memo, label: 備考 }
`;

const page = (): WizardPageDefinition =>
  parsePageYaml(YAML, { strict: true }) as WizardPageDefinition;

describe("ステップの出し分け", () => {
  it("読める（strict でも通る）", () => {
    const steps = page().steps;
    expect(steps[1].visibleWhen).toEqual({
      field: "kind",
      operator: "equals",
      value: "corp",
    });
    expect(steps[0].visibleWhen).toBeUndefined();
  });

  it("**歩くのは見えているステップだけ**（判定は1か所）", () => {
    const corp = visibleWizardSteps(page(), { kind: "corp" });
    expect(corp.map((one) => one.id)).toEqual(["basic", "billing", "confirm"]);
    const person = visibleWizardSteps(page(), { kind: "person" });
    expect(person.map((one) => one.id)).toEqual(["basic", "confirm"]);
    // 1件ずつでも同じ答え（送り側はこちらを使うこともある）。
    expect(isStepVisible(page().steps[1], { kind: "person" })).toBe(false);
  });

  it("**隠れたステップは検証しない**（そのステップだけを見るときも）", () => {
    const validator = new FormValidator();
    const billing = page().steps[1];
    // 個人なら、必須なのに鳴らない（誰にも見えていないので）。
    expect(
      validator.validate(wizardStepForm(billing), { kind: "person" }).errors,
    ).toEqual([]);
    // 法人なら鳴る。
    expect(
      validator
        .validate(wizardStepForm(billing), { kind: "corp" })
        .errors.map((one) => one.field),
    ).toEqual(["billTo"]);
  });

  it("**保存のときも飛ばす**（全部を1枚にした form でも同じ）", () => {
    const validator = new FormValidator();
    const whole = wizardForm(page());
    expect(
      validator.validate(whole, { kind: "person", memo: "" }).errors,
    ).toEqual([]);
    expect(
      validator
        .validate(whole, { kind: "corp" })
        .errors.map((one) => one.field),
    ).toEqual(["billTo"]);
  });

  it("条件はそのまま区画に載る（同じ規則を2つ書かない）", () => {
    const section = wizardForm(page()).sections[1];
    expect(section.visibleWhen).toEqual(page().steps[1].visibleWhen);
  });

  it("読み返しに出る（飛ばされるステップの理由が紙から消えない）", () => {
    const text = JSON.stringify(explainSource(YAML));
    expect(text).toContain("請求先");
    expect(text).toContain("区分");
    // 条件つきの題は、区画と同じ言い方で出る。
    expect(text).toMatch(/請求先[^"]*のときだけ/);
  });

  it("知らないキーは今までどおり弾く（穴を広げていない）", () => {
    const broken = YAML.replace("visibleWhen: {", "visibleIf: {");
    expect(() => parsePageYaml(broken, { strict: true })).toThrow(
      UnknownKeysError,
    );
  });
});
